src/rgw/rgw_lc_tier.cc

```diff
@@ -0,0 +1,1072 @@
+    return retcode;
+  }
+  /*
+   * If one of the following fails we cannot abort upload, as we cannot
+   * extract the upload id. If one of these fail it's very likely that that's
+   * the least of our problem.
+   */
+  RGWXMLDecoder::XMLParser parser;
+  if (!parser.init()) {
+    ldpp_dout(dpp, 0) << "ERROR: failed to initialize xml parser for parsing multipart init response from server" << dendl;
+    return -EIO;
+  }
+
+  if (!parser.parse(out_bl.c_str(), out_bl.length(), 1)) {
+    string str(out_bl.c_str(), out_bl.length());
+    ldpp_dout(dpp, 5) << "ERROR: failed to parse xml Completemultipart: " << str << dendl;
+    return -EIO;
+  }
+
+  try {
+    RGWXMLDecoder::decode_xml("CompleteMultipartUploadResult", result, &parser, true);
+  } catch (RGWXMLDecoder::err& err) {
+    string str(out_bl.c_str(), out_bl.length());
+    ldpp_dout(dpp, 5) << "ERROR: unexpected xml: " << str << dendl;
+    return -EIO;
+  }
+
+  ldpp_dout(dpp, 20) << "complete multipart result: location=" << result.location << " bucket=" << result.bucket << " key=" << result.key << " etag=" << result.etag << dendl;
+
+  return retcode;
+}
+
+int RGWLCCloudStreamAbortMultipartUpload::process() {
+  abort_mp.reset(new RGWLCCloudAbortMultipart(tier_ctx.dpp, tier_ctx.conn.get(), dest_obj, upload_id));
+
+  retcode = abort_mp->process();
+  if (retcode < 0) {
+    ldpp_dout(tier_ctx.dpp, 0) << "ERROR: failed to abort multipart upload dest obj=" << dest_obj << " upload_id=" << upload_id << " retcode=" << retcode << dendl;
+    /* ignore error, best effort */
+  }
+  /* remove status obj */
+  retcode = delete_sync_status(tier_ctx.dpp, tier_ctx.store, &status_obj);
+  if (retcode < 0) {
+    ldpp_dout(tier_ctx.dpp, 0) << "ERROR: failed to remove sync status obj obj=" << status_obj << " retcode=" << retcode << dendl;
+    // ignore error, best effort 
+  }
+  return 0;
+}
+
+int RGWLCCloudStreamObjToMultipart::process() {
+  obj_size = tier_ctx.o.meta.size;
+
+  target_bucket.name = tier_ctx.target_bucket_name;
+
+  target_obj_name = tier_ctx.bucket_info.bucket.name + "/" +
+    (*tier_ctx.obj)->get_name();
+  if (!tier_ctx.o.is_current()) {
+    target_obj_name += get_key_instance((*tier_ctx.obj)->get_key());
+  }
+  dest_obj.init(target_bucket, target_obj_name);
+
+  status_obj = rgw_raw_obj(tier_ctx.store->get_zone()->get_params().log_pool,
+      "lc_multipart_" + (*tier_ctx.obj)->get_oid());
+
+  retcode = read_sync_status(tier_ctx.dpp, tier_ctx.store, &status_obj, &status);
+
+  if (retcode < 0 && retcode != -ENOENT) {
+    ldpp_dout(tier_ctx.dpp, 0) << "ERROR: failed to read sync status of object " << src_obj << " retcode=" << retcode << dendl;
+    return retcode;
+  }
+
+  if (retcode >= 0) {
+    // check here that mtime and size did not change 
+    if (status.mtime != obj_properties.mtime || status.obj_size != obj_size ||
+        status.etag != obj_properties.etag) {
+      std::unique_ptr<RGWLCCloudStreamAbortMultipartUpload> abort_mp;
```

Comment:
ok, so the sync status object helps us identify incomplete transfers that need to be aborted. but outside of `mtime`/`obj_size`/`etag`/`upload_id`, that sync status stores a lot of extra stuff that it probably doesn't need to

it doesn't appear to be necessary to rewrite the sync status with `put_sync_status()` every time we update `cur_part`, `cur_ofs`, or `parts`; all of those variables could probably just be on the stack instead of encoding them inside the sync status object with `rgw_lc_multipart_upload_info`

if this `RGWLCCloudStreamAbortMultipartUpload` case is the only reason we're persisting the sync status object, then we should only write it once (immediately after `RGWLCCloudInitMultipart` succeeds)

Reply:
@cbodley .... yes now that all these calls are synchronous and not part of coroutines, there doesn't seem any need to store sync status object . Could you please confirm if it can be removed completely.

Reply:
i believe we still want it, so that we can abort previous attempts to upload:

> if this RGWLCCloudStreamAbortMultipartUpload case is the only reason we're persisting the sync status object, then we should only write it once (immediately after RGWLCCloudInitMultipart succeeds)

Reply:
okay thanks!